File: <100/055/055.py
def Lychrel_process(n: int) -> int:
    s = str(n)
    s = s[::-1]
    m = int(s)
    return n + m


def is_palindrome(n: int) -> bool:
    s = str(n)
    l = len(s)
    for i in range(l//2):
        if s[i] != s[l - 1 - i]:
            return False
    
    return True


def is_Lychrel(n: int, cnt: int = 0) -> bool:
    cnt += 1
    if 50 <= cnt:
        return True
    
    next_n = Lychrel_process(n)
    if is_palindrome(next_n):
        return False
    
    return is_Lychrel(next_n, cnt)


def main() -> int:
    upper_limit = 10**4
    filterd = list(filter(is_Lychrel, range(1, upper_limit)))
    return len(filterd)


if __name__ == "__main__":
    answer = main()
    print(answer)

File: <100/055/055.cpp
#include<iostream>
#include<algorithm>
using namespace std;


string add(string a, string b){
  string s = "";
  int n = a.size(), carry = 0;
  reverse(a.begin(), a.end()), reverse(b.begin(), b.end());

  for(int i = 0; i < n; ++i){
    int a_i = a[i] - '0', b_i = b[i] - '0';
    carry += a_i + b_i;

    s += (carry % 10) + '0';
    carry /= 10;
  }

  reverse(s.begin(), s.end());
  if(carry > 0){
    s = to_string(carry) + s;
  }
  return s;
}


string Lychrel_process(string s){
  string t = s;
  reverse(t.begin(), t.end());
  return add(s, t);
}


bool is_palindrome(string s){
  int n = s.size();
  for(int i = 0; i < n/2; ++i){
    if(s[i] != s[n-1-i]){
      return false;
    }
  }

  return true;
}


bool dfs(string s, int cnt=0){
  ++cnt;
  if(50 <= cnt){
    return true;
  }
  string next_s = Lychrel_process(s);

  if(is_palindrome(next_s)){
    return false;
  }

  return dfs(next_s, cnt);
}


bool is_Lychrel(int n){
  string s = to_string(n);
  return dfs(s);
}


int main(){
  int upper_limit = 10000, answer = 0;
  for(int n = 1; n < upper_limit; ++n)if(is_Lychrel(n)){
    answer += 1;
  }

  cout << answer << endl;
  return 0;
}
